applications/ray/kuberay-tpu-webhook/webhook_main_test.go

```diff
@@ -0,0 +1,595 @@
+package main
+
+import (
+	"testing"
+
+	rayv1 "github.com/ray-project/kuberay/ray-operator/apis/ray/v1"
+	"github.com/ray-project/kuberay/ray-operator/controllers/ray/utils"
+	admissionv1 "k8s.io/api/admission/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/utils/pointer"
+
+	"github.com/stretchr/testify/assert"
+	// "github.com/GoogleCloudPlatform/kuberay-tpu-webhook"
+)
+
+var (
+	namespaceStr                  string
+	instanceName                  string
+	groupNameStr                  string
+	headGroupNameStr              string
+	testPodAdmissionReviews       *admissionv1.AdmissionReview
+	testCPUPods                   []runtime.Object
+	testTPUPods                   []runtime.Object
+	testRayClusterAdmissionReview *admissionv1.AdmissionReview
+	testRayClusterNoTPUs          *rayv1.RayCluster
+	testRayClusterSingleHostTPU   *rayv1.RayCluster
+	testRayClusterMultiHostTPU    *rayv1.RayCluster
+	testServices                  []runtime.Object
+	workerSelector                labels.Selector
+	headNodeIP                    string
+	// sliceToWorkers				  map[slice][]worker
+	numOfHosts  int32
+	numReplicas int
+)
+
+func setupTest(t *testing.T) {
+	namespaceStr = "default"
+	instanceName = "raycluster-sample"
+	headNodeIP = "1.2.3.4"
+	groupNameStr = "workergroup"
+	headlessServiceSuffix = "headless-worker-svc"
+
+	// 1 CPU head pod + 1 worker - doesn't request TPUs
+	testCPUPods = []runtime.Object{
```

Comment:
Is there a way we could define constant yamls for something like "cpu worker" and "tpu worker", and then re-use these templates in subsequent tests? 

One option is to add a "testdata" directory where you could define these as yaml strings, which you could then parse here to reduce the repetition in test code.

Reply:
I think that's a good idea, for now I changed `testCPUPods` and `testTPUPods` to a cpu and tpu worker template whose values we can change within the test. In the next PR with the rest of the tests I think it'd be good to add a testdata directory (since these objects aren't being used yet in this PR).